scripts/src/main/resources/scripts/lib_script/import.groovy

```diff
@@ -61,12 +76,44 @@ class MyWorkbenchAdvisor extends org.eclipse.ui.application.WorkbenchAdvisor {
     if (!projectDirectory.canRead()) {
       throw new IllegalStateException("Cannot open project directory " + projectDirectoryName);
     }
-    SmartImportJob job = new SmartImportJob(projectDirectory, getOrCreateWorkingSets(workingSetNames), true, true);
-    Map proposals = job.getImportProposals(null);
-    job.setDirectoriesToImport(proposals.keySet());
-    job.setListener(new SysoutListener());
-    job.schedule();
-    job.join();
+    SmartImportJob importJob = new SmartImportJob(projectDirectory, getOrCreateWorkingSets(workingSetNames), true, true);
+    Map proposals = importJob.getImportProposals(null);
+    importJob.setDirectoriesToImport(proposals.keySet());
+    importJob.setListener(new SysoutListener());
+    importJob.run(null);
+    
+    MavenUpdateRequest request = new MavenUpdateRequest(project, isOffline, isForceUpdate);
+    configurationManager.updateProjectConfiguration(request, monitor);
+
+    waitForJobs();
+  }
+  
+  private void waitForJobs() {
+    //Bundle bundle = Platform.getBundle("org.eclipse.m2e.importer");
+    //Class c = bundle.loadClass("org.eclipse.m2e.importer.internal.MavenProjectConfigurator\$UpdateMavenConfigurationJob"); 
+    //Method m = c.getMethod("getInstance");
+    //Object o = m.invoke(null, null);
+    //o.join();
+    
+    // Wait for other jobs
+    List importantJobClassNames = Arrays.asList(
+          "org.eclipse.m2e.importer.internal.MavenProjectConfigurator\$UpdateMavenConfigurationJob",
+          "org.eclipse.m2e.core.internal.project.registry.ProjectRegistryRefreshJob",
+          "org.eclipse.core.internal.events.AutoBuildJob",
+      );
+      System.out.println("Wating for jobs to finish...");
+      IJobManager jobMan = Job.getJobManager();
+      boolean importantJobsRunning;
+      do {
+        importantJobsRunning = false;
+        System.out.println("Still running:");
+        Job[] jobs = jobMan.find(null);
+        for (Job job : jobs) {
+          importantJobsRunning |= (importantJobClassNames.contains(job.getClass().getName()));
+          System.out.println(job.toString() + ": " + job.getClass().getName());
+        }  
+        Thread.sleep(3000);      
+      } while (importantJobsRunning);
```

Comment:
I think it's a smart idea to wait for specific job names, but in my opinion, that also makes the maintenance harder in case there are some classes renamed/packages renamed of the plugins, which is hard to detect.
Also eclipse jobs are triggered after another and even by other jobs, so it might be, that any of the jobs might even just being triggered when actually being checked for the running status. This might be some very rare cases. But nevertheless.

My proposal wouldbe to simply wait until the jobmanager has processed all jobs: `Job.getJobManager().isIdle()`. This might take a little longer, but you could be sure, that also future "jobs" to fix some other issues or bring some new functionality into eclipse will be executed properly.

Also this might fix the refresh problem, you faced as maybe some finishing jobs are taking care of workspace polishing?

Reply:
I tried `Job.getJobManager().isIdle()` but it seems that it is never `true` meaning there is always some job running.
Maybe I can check what eclipse is normally doing to assure a clean shutdown. Currently we call `PlatformUI.getWorkbench().close();` maybe there is something more todo...

Reply:
Ok, that's strange, we are using this in our swtbot tests all the time: https://github.com/devonfw/cobigen/blob/master/cobigen-eclipse/cobigen-eclipse-test/src/main/java/com/devonfw/cobigen/eclipse/test/common/swtbot/AllJobsAreFinished.java 

It has been turned out, that this is the most stable solution, but it honestly takes a little longer. Eclipse seems to do quite some stuff. Our timeouts for that are at 10s, working stable to wait for eclipse. But obviously, it also depends on the number of plugins you installed what the list of tasks is. :(